### ISSUE TITLE
INCONSISTENCY: Private loader on Linux inconsistently initializes thread local variables

### ISSUE THREAD

**Issue by edmcman** [DynamoRIO/dynamorio#3973]
**Describe the bug**
Thread local variables in client code are sometimes initialized to `0xabababab` instead of the value they are defined to be.

**To Reproduce**

**concrete_eval.cpp**
``` c++
#include "dr_api.h"

__thread int foo = 42;

static void wtf (void* ) {
  dr_printf ("foo = %x\n", foo);
}

DR_EXPORT void
dr_client_main(client_id_t id, int argc, const char *argv[])
{

  dr_printf ("main = %x\n", foo);

  dr_create_client_thread (wtf, NULL);
  dr_create_client_thread (wtf, NULL);
  dr_create_client_thread (wtf, NULL);

}
```

**CMakeLists.txt**
```
set (CMAKE_CXX_STANDARD 17)
add_library(concrete_eval SHARED concrete_eval.cpp)

find_package(DynamoRIO)
if (NOT DynamoRIO_FOUND)
  message(FATAL_ERROR "DynamoRIO package required to build")
endif(NOT DynamoRIO_FOUND)
configure_DynamoRIO_client(concrete_eval)
```

```/media/data-linux/dynamorio/build/bin64/drrun -debug -c libconcrete_eval.so -- ls
<Starting application /bin/ls (6020)>
<Initial options = -no_dynamic_options -client_lib '/media/data-linux/concolic-dynamorio-experiment/build/libconcrete_eval.so;0;' -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
<Paste into GDB to debug DynamoRIO clients:
set confirm off
add-symbol-file '/media/data-linux/concolic-dynamorio-experiment/build/libconcrete_eval.so' 0x00007fffb3d1d7f0
add-symbol-file '/media/data-linux/dynamorio/build/lib64/debug/libdynamorio.so' 0x0000000071040c58
add-symbol-file '/lib64/ld-linux-x86-64.so.2' 0x00007ffff7aeaf10
>
main = 2a
foo = abababab
foo = 2a
foo = 2a
<(1+x) Handling our fault in a TRY at 0x000000007129c505>
build.ninja  CMakeCache.txt  CMakeFiles  cmake_install.cmake  concrete_eval.ldscript  libconcrete_eval.so  rules.ninja
<Stopping application /bin/ls (6020)>
```

Command: `/media/data-linux/dynamorio/build/bin64/drrun -debug -c libconcrete_eval.so -- ls`
yields
```
<Starting application /bin/ls (7604)>
<Initial options = -no_dynamic_options -client_lib '/media/data-linux/concolic-dynamorio-experiment/build/libconcrete_eval.so;0;' -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
<Paste into GDB to debug DynamoRIO clients:
set confirm off
add-symbol-file '/media/data-linux/concolic-dynamorio-experiment/build/libconcrete_eval.so' 0x00007fffb3d1d7f0
add-symbol-file '/media/data-linux/dynamorio/build/lib64/debug/libdynamorio.so' 0x0000000071040c58
add-symbol-file '/lib64/ld-linux-x86-64.so.2' 0x00007ffff7aeaf10
>
main = 2a
foo = abababab
foo = 2a
foo = 2a
<(1+x) Handling our fault in a TRY at 0x000000007129c505>
build.ninja  CMakeCache.txt  CMakeFiles  cmake_install.cmake  concrete_eval.ldscript  libconcrete_eval.so  rules.ninja
<Stopping application /bin/ls (7604)>
```

Please also answer these questions:
 - What happens when you run without any client?
`ls` runs normally.
 - What happens when you run with debug build ("-debug" flag to drrun/drconfig/drinject)?
I am using a debug build.

**Expected behavior**
I expected the program to print:
```
main = 2a
foo = 2a
foo = 2a
foo = 2a
```

**Versions**
 - What version of DynamoRIO are you using?
eab941b43e5aef8447aeb924883c5552aacd9a3e
 - Does the latest build from
https://github.com/DynamoRIO/dynamorio/wiki/Latest-Build solve the problem?
No.
- What operating system version are you running on? ("Windows 10" is *not* sufficient: give the release number.)
Ubuntu 18.04.03
- Is your application 32-bit or 64-bit?
64-bit

**Other notes**

I have only observed the first (explicitly) created thread having the problem.  I have also observed some non-determinism of the problem, though minimizing the client helped reduce that.

```
for i in $(seq 100); do /media/data-linux/dynamorio/build/bin64/drrun -debug -c libconcrete_eval.so -- ls; done |& fgrep foo | sort | uniq -c
    202 foo = 2a
     98 foo = abababab
```

In other words, the variable is initialized correctly 2/100 times.

**Comment by edmcman** [comment]
The problem that led me to this was an intermittent crash when calling `__ctype_init` (#3972).  The problem appears to be that `__ctype_init` accesses a thread local variable that points to a locale object.  When that variable is not initialized properly, `__ctype_init` crashes.

**Comment by derekbruening** [comment]
The private library TLS is implemented in core/unix/loader_linux.c if you would like to take a shot at finding and fixing the bug.

**Comment by edmcman** [comment]
I'll take a look...

Do you know where the `0xabababab` pattern comes from?

**Comment by edmcman** [comment]
So the first time `privload_copy_tls_block` is called for a new thread, the destination does not match the address that the C code sees.  But for subsequent threads it does...

```
hmm abababab 0x00007fff33e0d24c
Server thread
hmm 2a 0x00007fff33e9524c
```

```
privload_copy_tls_block: copying ELF TLS from 0x00007fffb3fcc01c to 0x00007fff33e7124c block 4 image 4
0x00007fffb3fcc01c 2a 00 00 00 
0x00007fff33e7124c 2a 00 00 00 
```

```
privload_copy_tls_block: copying ELF TLS from 0x00007fffb3fcc01c to 0x00007fff33e9524c block 4 image 4
0x00007fffb3fcc01c 2a 00 00 00 
0x00007fff33e9524c 2a 00 00 00
```

**Comment by derekbruening** [comment]
> I'll take a look...

Thanks!

> Do you know where the `0xabababab` pattern comes from?

Yes, DR debug build fills in allocated but uninitialized memory with 0xab (and freed memory with 0xcd).

**Comment by edmcman** [comment]
``` c++
    /* We initialize the primary thread's ELF TLS in privload_mod_tls_init()
     * after finalizing the module load (dependent libs not loaded yet here).
     * For subsequent threads we walk the module list here.
     */
    LOG(GLOBAL, LOG_LOADER, 2, "dynamo_initialized = %d\n", dynamo_initialized);
    if (dynamo_initialized) {
```

This line is reached twice with `dynamo_initialized == 0`.  I suspect there is a race condition with the flag being set and the new thread being created.

**Comment by edmcman** [comment]
A fix pull request is in #3976.

**Comment by derekbruening** [comment]
So a key aspect here is that this is a *client thread*, not an app thread.  See discussion in the PR #3976 on possibly delaying client thread init to solve this.
